src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -97,7 +97,19 @@ public sealed class NpgsqlDataReader : DbDataReader, IDbColumnSchemaGenerator
         /// Mostly useful for a sequential mode, when the row is already in the buffer.
         /// Should always be true for the non-sequential mode.
         /// </summary>
-        bool CanConsumeRowNonSequential => _dataMsgEnd - Buffer.ReadPosition <= Buffer.ReadBytesLeft;
+        bool CanConsumeRowNonSequentially
+        {
+            get
+            {
+                var result = _dataMsgEnd - Buffer.ReadPosition <= Buffer.ReadBytesLeft;
+                if (!result)
```

Comment:
I'm pretty sure this gets optimized correctly in Release (empty if), but maybe just (`Debug.Assert(result || _isSequential)`)?